File: source/ReferralNetwork.java
import java.util.Collections;
import java.util.HashMap;
import java.util.HashSet;
import java.util.LinkedList;
import java.util.Map;
import java.util.Queue;
import java.util.Set;

/**
 * Implements the core data structure for a referral network. [cite: 17, 20]
 * This class stores users and their directed referral relationships,
 * enforcing rules such as no self-referrals, unique referrers for each candidate,
 * and maintaining an acyclic graph structure. [cite: 21, 26, 27, 28]
 */
public class ReferralNetwork {

    // Adjacency list to represent the graph: Referrer -> Set of Candidates
    private final Map<String, Set<String>> graph;

    // Map to enforce unique referrer constraint: Candidate -> Referrer
    private final Map<String, String> referredCandidates;

    public ReferralNetwork() {
        this.graph = new HashMap<>();
        this.referredCandidates = new HashMap<>();
    }

    /**
     * Adds a new, directed referral link from a referrer to a candidate. [cite: 22]
     * This operation is transactional: it succeeds only if all constraints are met.
     *
     * @param referrer  The ID of the user making the referral.
     * @param candidate The ID of the user being referred.
     * @return true if the referral was added successfully, false otherwise.
     */
    public boolean addReferral(String referrer, String candidate) {
        // Constraint 1: No Self-Referrals [cite: 26]
        if (referrer.equals(candidate)) {
            System.err.println("Error: A user cannot refer themselves.");
            return false;
        }

        // Constraint 2: A candidate can only be referred by one user. [cite: 27]
        if (referredCandidates.containsKey(candidate)) {
            System.err.println("Error: Candidate '" + candidate + "' has already been referred by '" + referredCandidates.get(candidate) + "'.");
            return false;
        }

        // Constraint 3: The graph must remain acyclic. [cite: 28]
        // To check, we verify if a path already exists from the candidate to the referrer.
        // If it does, adding an edge from referrer to candidate would create a cycle.
        if (pathExists(candidate, referrer)) {
            System.err.println("Error: Adding this referral would create a cycle in the network.");
            return false;
        }

        // If all constraints pass, add the referral
        graph.computeIfAbsent(referrer, k -> new HashSet<>()).add(candidate);
        referredCandidates.put(candidate, referrer);

        return true;
    }

    /**
     * Queries for a user's direct referrals. [cite: 23]
     *
     * @param user The ID of the user.
     * @return An unmodifiable set of the user's direct referrals. Returns an empty set if the user has no referrals.
     */
    public Set<String> getDirectReferrals(String user) {
        return Collections.unmodifiableSet(graph.getOrDefault(user, Collections.emptySet()));
    }

    /**
     * Checks if a directed path exists from a start node to an end node using Breadth-First Search (BFS).
     *
     * @param startNode The starting user ID.
     * @param endNode   The target user ID.
     * @return true if a path exists, false otherwise.
     */
    private boolean pathExists(String startNode, String endNode) {
        if (!graph.containsKey(startNode)) {
            return false;
        }

        Queue<String> queue = new LinkedList<>();
        Set<String> visited = new HashSet<>();

        queue.add(startNode);
        visited.add(startNode);

        while (!queue.isEmpty()) {
            String currentUser = queue.poll();

            if (currentUser.equals(endNode)) {
                return true;
            }

            for (String neighbor : getDirectReferrals(currentUser)) {
                if (!visited.contains(neighbor)) {
                    visited.add(neighbor);
                    queue.add(neighbor);
                }
            }
        }
        return false;
    }
}

File: frontend/src/ReferralNetwork.js
/**
 * Referral Network - Core Graph Implementation
 * 
 * This class implements a directed acyclic graph (DAG) to manage referral relationships.
 * Each user can refer multiple candidates, but each candidate can only have one referrer.
 */

export class ReferralNetwork {
  constructor() {
    // Store referral relationships: referrer -> Set of candidates
    this.referrals = new Map();
    // Reverse mapping: candidate -> referrer (for unique referrer constraint)
    this.referrers = new Map();
    // Set of all users in the network
    this.users = new Set();
  }

  /**
   * Add a user to the network
   * @param {string} userId - Unique identifier for the user
   */
  addUser(userId) {
    if (!userId || typeof userId !== 'string') {
      throw new Error('User ID must be a non-empty string');
    }
    
    this.users.add(userId);
    if (!this.referrals.has(userId)) {
      this.referrals.set(userId, new Set());
    }
  }

  /**
   * Add a referral relationship from referrer to candidate
   * @param {string} referrerId - User making the referral
   * @param {string} candidateId - User being referred
   * @returns {boolean} - Success status
   */
  addReferral(referrerId, candidateId) {
    // Validation checks
    if (!referrerId || !candidateId) {
      throw new Error('Both referrer and candidate IDs are required');
    }

    if (referrerId === candidateId) {
      throw new Error('Self-referrals are not allowed');
    }

    // Add users if they don't exist
    this.addUser(referrerId);
    this.addUser(candidateId);

    // Check if candidate already has a referrer
    if (this.referrers.has(candidateId)) {
      throw new Error(`Candidate ${candidateId} already has a referrer: ${this.referrers.get(candidateId)}`);
    }

    // Check for cycle creation using DFS
    if (this.wouldCreateCycle(referrerId, candidateId)) {
      throw new Error('Adding this referral would create a cycle');
    }

    // Add the referral relationship
    this.referrals.get(referrerId).add(candidateId);
    this.referrers.set(candidateId, referrerId);

    return true;
  }

  /**
   * Check if adding a referral would create a cycle
   * @param {string} referrerId - Proposed referrer
   * @param {string} candidateId - Proposed candidate
   * @returns {boolean} - True if cycle would be created
   */
  wouldCreateCycle(referrerId, candidateId) {
    // If candidate can reach referrer through existing paths, adding this edge creates a cycle
    return this.canReach(candidateId, referrerId);
  }

  /**
   * Check if one user can reach another through referral chain
   * @param {string} fromUser - Starting user
   * @param {string} toUser - Target user
   * @returns {boolean} - True if path exists
   */
  canReach(fromUser, toUser) {
    if (fromUser === toUser) return true;
    
    const visited = new Set();
    const stack = [fromUser];

    while (stack.length > 0) {
      const current = stack.pop();
      if (visited.has(current)) continue;
      visited.add(current);

      const directReferrals = this.referrals.get(current);
      if (directReferrals) {
        for (const referral of directReferrals) {
          if (referral === toUser) return true;
          stack.push(referral);
        }
      }
    }

    return false;
  }

  /**
   * Get direct referrals for a user
   * @param {string} userId - User ID
   * @returns {Set} - Set of direct referrals
   */
  getDirectReferrals(userId) {
    return this.referrals.get(userId) || new Set();
  }

  /**
   * Get the referrer of a user (if any)
   * @param {string} userId - User ID
   * @returns {string|null} - Referrer ID or null
   */
  getReferrer(userId) {
    return this.referrers.get(userId) || null;
  }

  /**
   * Get all users in the network
   * @returns {Set} - Set of all user IDs
   */
  getAllUsers() {
    return new Set(this.users);
  }

  /**
   * Get network statistics
   * @returns {Object} - Network statistics
   */
  getStats() {
    return {
      totalUsers: this.users.size,
      totalReferrals: Array.from(this.referrals.values())
        .reduce((sum, referrals) => sum + referrals.size, 0),
      usersWithReferrals: Array.from(this.referrals.values())
        .filter(referrals => referrals.size > 0).length
    };
  }
}
